trio/_timeouts.py

```diff
@@ -120,7 +124,18 @@ def fail_at(deadline):
         raise TooSlowError
 
 
-def fail_after(seconds):
+# workaround for PyCharm not being able to infer return type from @contextmanager
+# see https://youtrack.jetbrains.com/issue/PY-36444/PyCharm-doesnt-infer-types-when-using-contextlib.contextmanager-decorator
+if TYPE_CHECKING:
+
+    def fail_at(deadline: float) -> AbstractContextManager[trio.CancelScope]:
+        return _fail_at(deadline)
+
+else:
+    fail_at = _fail_at
```

Comment:
The suggested workaround in https://youtrack.jetbrains.com/issue/PY-36444/PyCharm-doesnt-infer-types-when-using-contextlib.contextmanager-decorator is ugly as shit:

```python
def successful_workaround_using_wrapper_function_annotated_returning_context_manager():
    """PyCharm autocompletion works and mypy is also happy, however this is an ugly workaround"""

    def wrapper() -> ContextManager[str]:
        @contextmanager
        def generator_function():
            yield "some value"

        return generator_function()

    with wrapper() as value:
        print(value.upper())  # PyCharm autocompletion works
```

I think this one is much better, it allows both pycharm autocompletion as well as typechecking of the original function, doesn't incur any runtime penalt, and is relatively small and unobtrusive

Reply:
I'd suggest setting `__name__` to `fail_at`, or just move the function under the `else:` block. The type checking version can just have a body of `...`.

Reply:
ah, we're losing the docstring

Reply:
Another way to do it perhaps:
```py
def fail_at(...) -> ContextManager[trio.CancelScope]: # type: ignore
    ...

if not TYPE_CHECKING:
    fail_at = contextmanager(fail_at)
```
Just hide the decorator part from the type checker, though we need an ignore comment since Mypy would rightly complain that the return type is wrong.

Reply:
> I'd suggest setting `__name__` to `fail_at`, or just move the function under the `else:` block. The type checking version can just have a body of `...`.

I was thinking the docstring need to be accessible both during runtime and during static analysis

Reply:
I also came up with:
```python
# workaround for PyCharm not being able to infer return type from @contextmanager
# see https://youtrack.jetbrains.com/issue/PY-36444/PyCharm-doesnt-infer-types-when-using-contextlib.contextmanager-decorator
@contextmanager
def _fail_at(deadline: float) -> Iterator[trio.CancelScope]:
    with move_on_at(deadline) as scope:
        yield scope
    if scope.cancelled_caught:
        raise TooSlowError

def fail_at(deadline: float) -> AbstractContextManager[trio.CancelScope]:
    """Creates a cancel scope with the given deadline, and raises an error if it
    ...

    """
    return _fail_at(deadline)
```

Reply:
> Another way to do it perhaps:
> 
> ```python
> def fail_at(...) -> ContextManager[trio.CancelScope]: # type: ignore
>     ...
> 
> if not TYPE_CHECKING:
>     fail_at = contextmanager(fail_at)
> ```
> 
> Just hide the decorator part from the type checker, though we need an ignore comment since Mypy would rightly complain that the return type is wrong.

this one seems pretty clean though, and doesn't introduce another frame like the above